models/ed/R/met2model.ED2.R

```diff
@@ -30,297 +30,299 @@
-      doy <- dtmp
-      hr <- rep(NA,length(dtmp))
-    } else {
-      yr <- c(yr,ytmp)
-      doy <- c(doy,dtmp)
-      hr <- c(hr,rep(NA,length(dtmp)))
+    
+    flon <- try(ncvar_get(nc, "longitude"), silent = TRUE)
+    if (!is.numeric(flon)) 
+      flat <- nc$dim[[2]]$vals[1]
+    if (is.na(lon)) {
+      lon <- flon
+    } else if (lon != flon) {
+      logger.warn("Longitude does not match that of file", lon, "!=", flon)
     }
-    rng <- length(doy) - length(ytmp):1 + 1
-    if(!all(rng>=0)){
-      skip = TRUE
-      logger.warn(paste(year,"is not a complete year and will not be included"))
-      break
+    
+    ## determine GMT adjustment lst <- site$LST_shift[which(site$acro == froot)]
+    
+    ## extract variables
+    lat  <- eval(parse(text = lat))
+    lon  <- eval(parse(text = lon))
+    sec  <- nc$dim$time$vals
+    Tair <- ncvar_get(nc, "air_temperature")
+    Qair <- ncvar_get(nc, "specific_humidity")  #humidity (kg/kg)
+    U    <- ncvar_get(nc, "eastward_wind")
+    V    <- ncvar_get(nc, "northward_wind")
+    Rain <- ncvar_get(nc, "precipitation_flux")
+    pres <- ncvar_get(nc, "air_pressure")
+    SW   <- ncvar_get(nc, "surface_downwelling_shortwave_flux_in_air")
+    LW   <- ncvar_get(nc, "surface_downwelling_longwave_flux_in_air")
+    CO2  <- try(ncvar_get(nc, "mole_fraction_of_carbon_dioxide_in_air"), silent = TRUE)
+    
+    useCO2 <- is.numeric(CO2)
+    
+    ## convert time to seconds
+    sec <- udunits2::ud.convert(sec, unlist(strsplit(nc$dim$time$units, " "))[1], "seconds")
+    
+    nc_close(nc)
+    
+    dt <- ifelse(leap_year(year) == TRUE, 
+                 366 * 24 * 60 * 60 / length(sec), # leap year
+                 365 * 24 * 60 * 60 / length(sec)) # non-leap year
+    
+    toff <- -as.numeric(lst) * 3600 / dt
+    
+    ## buffer to get to GMT
+    slen <- length(SW)
+    Tair <- c(rep(Tair[1], toff), Tair)[1:slen]
+    Qair <- c(rep(Qair[1], toff), Qair)[1:slen]
+    U    <- c(rep(U[1], toff), U)[1:slen]
+    V    <- c(rep(V[1], toff), V)[1:slen]
+    Rain <- c(rep(Rain[1], toff), Rain)[1:slen]
+    pres <- c(rep(pres[1], toff), pres)[1:slen]
+    SW   <- c(rep(SW[1], toff), SW)[1:slen]
+    LW   <- c(rep(LW[1], toff), LW)[1:slen]
+    if (useCO2) {
+      CO2 <- c(rep(CO2[1], toff), CO2)[1:slen]
     }
-    asec[rng] <- asec[rng] - asec[rng[1]]
-    hr[rng] <- (asec[rng] - (dtmp-1)*86400)/86400*24
-  }
-  mo<-day2mo(yr,doy)
-  if(length(yr) < length(sec)){
-    rng <- (length(yr)+1):length(sec)
-    if(!all(rng>=0)){
-      skip = TRUE
-      logger.warn(paste(year,"is not a complete year and will not be included"))
-      break
+    
+    ## build time variables (year, month, day of year)
+    skip <- FALSE
+    nyr  <- floor(length(sec) / 86400 / 365 * dt)
+    yr   <- NULL
+    doy  <- NULL
+    hr   <- NULL
+    asec <- sec
+    for (y in (year + 1):(nyr - 1)) {
+      ytmp <- rep(y, 365 * 86400 / dt)
+      dtmp <- rep(1:365, each = 86400 / dt)
+      if (y%%4 == 0) {
```

Comment:
To fix this - can I go through the `models` scripts and have them all use `lubridate::leap_year`?